qubes-trust-daemon.cpp

```diff
@@ -0,0 +1,355 @@
+        }
+
+        // Add NULL terminator to signal end of argument list
+        qvm_argv[arg_index] = (char*) NULL;
+
+        // Once we have a list of MAX_ARG_LEN args,
+        // fork and attempt to call qvm-file-trust
+        switch (child_pid=fork()) {
+            case 0:
+                // We're the child, call qvm-file-trust
+                execv("/usr/bin/qvm-file-trust", (char**) qvm_argv);
+
+                // Unreachable if no error
+                perror("execl qvm-file-trust");
+                exit(1);
+            case -1:
+                // Fork failed
+                perror("fork failed");
+                exit(1);
+            default:
+                // Fork succeeded, and we got our pid, wait until child exits
+                if (waitpid(child_pid, &exit_code, 0) == -1) {
+                    perror("wait for qvm-file-trust failed");
+                    exit(1);
+                }
+        }
+    }
+
+    untrusted_buffer.clear();
+}
+
+/*
+ * Places an inotify_watch on the directory and all subdirectories
+ * Should call above method after compiling a list of all files to set as untrusted
+ */
+int watch_dir(const char *filepath, const struct stat *info,
+              const int typeflag, struct FTW *pathinfo) {
+    // Watch directories, set files as untrusted
+	struct stat s;
+	if(stat(filepath, &s) == 0) {
+		if(!(s.st_mode & S_IFDIR)) {
+            // File, set as untrusted
+            untrusted_buffer.insert(filepath);
+			return 0;
+		}
+	}
+	else {
+		// Error reading
+		return 1;
+	}
+
+    std::cout << "Placing watch on " << filepath
+        << " and subdirectories" << std::endl;
+
+    // Add watch to starting directory
+    int wd = inotify_add_watch(watch_fd, filepath, 
+		IN_CREATE | IN_MODIFY | IN_DELETE | IN_MOVED_TO | IN_MOVED_FROM);
+    if (wd == -1) {
+        printf("Couldn't add watch to %s\n", filepath);
+    } else {
+        printf("%d Watching:: %s\n", wd, filepath);
+        
+        // Add watch descriptor and filepath to global watchtable
+        std::string filepath_string = filepath;
+        std::pair<int, std::string> watch_pair(wd, filepath_string);
+        watch_table.insert(watch_pair);
+    }
+
+	return 0;
+}
+
+/*
+ * Walks a filepath for all files and directories contained within
+ */
+int place_watch_on_dir_and_subdirs(const char* const filepath) {
+	int result;
+
+    // Check for incorrect path
+    if (filepath == NULL || *filepath == '\0') {
+		return errno = EINVAL;
+    }
+
+	// Run watch_dir on directory and subdirectories
+	result = nftw(filepath, watch_dir, USE_FDS, FTW_PHYS);
+
+	if (result >= 0) {
+		errno = result;
+	}
+
+	return errno;
+}
+
+/* 
+ * Watches directories and acts on various spawned inotify events
+ */
+void keep_watch_on_dirs(const int fd) {
+    while(1) {
+        char buffer[BUF_LEN];
+        int length = read(fd, buffer, BUF_LEN);
+        int i = 0;
+
+        if (length < 0) {
+            perror("read");
+        }
```

Comment:
what about read returning 0 (EOF)?

Reply:
Thus no need to read from it, right? So instead it should check for `(length <= 0)`?

Reply:
Yes.

Reply:
Fixed in 790c49b